Spark docs in datasets libraries (#1392)
huggingface/hub-docs: docs/hub/datasets-spark.md

        where op is [==, =, >, >=, <, <=, !=, in, not in]
        The innermost tuples are transposed into a set of filters applied
        through an `AND` operation.
        The outer list combines these sets of filters through an `OR`
        operation.
        A single list of tuples can also be used, meaning that no `OR`
        operation between set of filters is to be conducted.

    **kwargs
        Any additional kwargs are passed to pyarrow.parquet.read_table.

    Returns
    -------
    DataFrame
        DataFrame based on parquet file.

    Examples
    --------
    >>> path = "hf://datasets/username/dataset/data.parquet"
    >>> pd.DataFrame({"foo": range(5), "bar": range(5, 10)}).to_parquet(path)
    >>> read_parquet(path).show()
    +---+---+
    |foo|bar|
    +---+---+
    |  0|  5|
    |  1|  6|
    |  2|  7|
    |  3|  8|
    |  4|  9|
    +---+---+
    >>> read_parquet(path, columns=["bar"]).show()
    +---+
    |bar|
    +---+
    |  5|
    |  6|
    |  7|
    |  8|
    |  9|
    +---+
    >>> sel = [("foo", ">", 2)]
    >>> read_parquet(path, filters=sel).show()
    +---+---+
    |foo|bar|
    +---+---+
    |  3|  8|
    |  4|  9|
    +---+---+
    """

>     filesystem: HfFileSystem = kwargs.pop("filesystem", HfFileSystem(**kwargs.pop("storage_options", {})))

Reviewer: ```suggestion
    filesystem: HfFileSystem = kwargs.pop("filesystem") if "filesystem" in kwargs else HfFileSystem(**kwargs.pop("storage_options", {}))
```

(nit) doing so avoid to instantiate a `HfFileSystem` obj for nothing if `filesystem` already defined.

    paths = filesystem.glob(path)
    if not paths:
        raise FileNotFoundError(f"Counldn't find any file at {path}")
    df = spark.createDataFrame([{"path": path} for path in paths])
    arrow_schema = pq.read_schema(filesystem.open(paths[0]))
    schema = pa.schema(
        [
            field for field in arrow_schema
            if (columns is None or field.name in columns)
        ],
        metadata=arrow_schema.metadata
    )
    return df.mapInArrow(
        partial(
            _read,
            columns=columns,
            filters=filters,
            filesystem=filesystem,
            schema=schema,
            **kwargs
        ),
        from_arrow_schema(schema)
    )
```

Here is how we can use this on the [BAAI/Infinity-Instruct](https://huggingface.co/datasets/BAAI/Infinity-Instruct) dataset.

> It is a gated repository, users have to accept the terms of use berfore accessing it.

Reviewer: ```suggestion
It is a gated repository, users have to accept the terms of use before accessing it.
```

| 14|        [{human, In react...|     |        en|                flan|
| 15|        [{human, Write Py...|     |        en|      code_exercises|
| 16|        [{human, Find the...|     |        en|            MetaMath|
| 17|        [{human, Three of...|     |        en|            MetaMath|
| 18|        [{human, Chandra ...|     |        en|            MetaMath|
| 19|[{human, 用经济学知识分析...|     |     zh-cn|          Subjective|
+---+----------------------------+-----+----------+--------------------+
```

To compute the number of dialogues per language we run this code.
The `columns` argument is useful to only load the data we need, since PySpark doesn't enable predicate push-down in this case.
There is also a `filters` argument to only load data with values within a certain range.

```python
>>> df_langdetect_only = read_parquet("hf://datasets/BAAI/Infinity-Instruct/7M/*.parquet", columns=["langdetect"])
>>> df_langdetect_only.groupBy("langdetect").count().show()
+----------+-------+                                                            
|langdetect|  count|
+----------+-------+
|        en|6697793|
|     zh-cn| 751313|
+----------+-------+
```

> To filter the dataset and only keep dialogues in chinese:

Reviewer: ```suggestion
To filter the dataset and only keep dialogues in Chinese:
```

```python

> >>> sel = sel = [("langdetect", "=", "zh-cn")]
> >>> df_chinese_only = read_parquet("hf://datasets/BAAI/Infinity-Instruct/7M/*.parquet", filters=sel)

Reviewer: ```suggestion
>>> criteria = [("langdetect", "=", "zh-cn")]
>>> df_chinese_only = read_parquet("hf://datasets/BAAI/Infinity-Instruct/7M/*.parquet", filters=criteria)
```

|  9|[{human, 因果联系原则是法...|     |     zh-cn|Subjective|
| 19|[{human, 用经济学知识分析...|     |     zh-cn|Subjective|
| 38| [{human, 某个考试共有A、...|     |     zh-cn|Subjective|
| 39|[{human, 撰写一篇关于斐波...|     |     zh-cn|Subjective|
| 57|[{human, 总结世界历史上的...|     |     zh-cn|Subjective|
| 61|[{human, 生成一则广告词。...|     |     zh-cn|Subjective|
| 66|[{human, 描述一个有效的团...|     |     zh-cn|Subjective|
| 94|[{human, 如果比利和蒂芙尼...|     |     zh-cn|Subjective|
|102|[{human, 生成一句英文名言...|     |     zh-cn|Subjective|
|106|[{human, 写一封感谢信，感...|     |     zh-cn|Subjective|
|118| [{human, 生成一个故事。}...|     |     zh-cn|Subjective|
|174|[{human, 高胆固醇水平的后...|     |     zh-cn|Subjective|
|180|[{human, 基于以下角色信息...|     |     zh-cn|Subjective|
|192|[{human, 请写一篇文章，概...|     |     zh-cn|Subjective|
|221|[{human, 以诗歌形式表达对...|     |     zh-cn|Subjective|
|228|[{human, 根据给定的指令，...|     |     zh-cn|Subjective|
|236|[{human, 打开一个新的生成...|     |     zh-cn|Subjective|
|260|[{human, 生成一个有关未来...|     |     zh-cn|Subjective|
|268|[{human, 如果有一定数量的...|     |     zh-cn|Subjective|
|273| [{human, 题目：小明有5个...|     |     zh-cn|Subjective|
+---+----------------------------+-----+----------+----------+
```

## Write

> We also provide a helper function to write datasets in a distributed manner to your Hugging Face account.

Reviewer: ```suggestion
We also provide a helper function to write datasets in a distributed manner to a Hugging Face repository.
```

In particular it uses the `preupload_lfs_files` utility to upload Parquet files in parallel in a distributed manner, and only commit the files once they're all uploaded:

```python
import math
import pickle
import tempfile
from functools import partial
from typing import Iterator, Optional
from uuid import uuid4

import pyarrow as pa
import pyarrow.parquet as pq
from huggingface_hub import CommitOperationAdd, CommitOperationCopy, CommitOperationDelete, HfFileSystem
from pyspark.sql.dataframe import DataFrame
from pyspark.sql.pandas.types import from_arrow_schema, to_arrow_schema

def _write_single(iterator: Iterator[pa.RecordBatch], path: str, schema: pa.Schema, filesystem: HfFileSystem, row_group_size: Optional[int] = None, **kwargs) -> Iterator[pa.RecordBatch]:
    resolved_path = filesystem.resolve_path(path)
    with tempfile.NamedTemporaryFile(suffix=".parquet") as temp_file:
        with pq.ParquetWriter(temp_file.name, schema=schema, **kwargs) as writer:
            for batch in iterator:
                writer.write_batch(batch, row_group_size=row_group_size)

>         filesystem._api.upload_file(

Reviewer: Can't we use a `filesystem.open("w").write(...).flush()` or similar? Feels weird to introduce `filesystem` but to rely on `HfApi` in the end.

Same for `_rename` / `_preupload` / `_commit` where we can't really do without `HfApi`. What feels weird to me is that we use `filesystem` only for `filesystem.resolve_path(path)`. In the end, we could also just define a method

```py
def write_parquet(df: DataFrame, repo_id: str, path_in_repo: str, *, repo_type: Optional[str] = None, revision: Optional[str] = None, **kwargs) -> None:
```

Author: The API choice is mostly to fit the pyspark API which uses urls like s3://

Then it would be cool to use the filesystem only, but I don't think the preupload_lfs logic is available (it's the only way to do parallel uploads afaik)

    --------
    >>> spark.createDataFrame(pd.DataFrame({"foo": range(5), "bar": range(5, 10)}))
    >>> # Save to one file
    >>> write_parquet(df, "hf://datasets/username/dataset/data.parquet")
    >>> # OR save to a directory (possibly in many files)
    >>> write_parquet(df, "hf://datasets/username/dataset")
    """
    filesystem: HfFileSystem = kwargs.pop("filesystem", HfFileSystem(**kwargs.pop("storage_options", {})))
    if path.endswith(".parquet") or path.endswith(".pq"):
        df.coalesce(1).mapInArrow(
            partial(_write_single, path=path, schema=to_arrow_schema(df.schema), filesystem=filesystem, **kwargs),
            from_arrow_schema(pa.schema({"path": pa.binary()})),
        ).collect()
    else:
        df.mapInArrow(
            partial(_preupload, path=path, schema=to_arrow_schema(df.schema), filesystem=filesystem, **kwargs),
            from_arrow_schema(pa.schema({"addition": pa.binary()})),
        ).coalesce(1).mapInArrow(
            partial(_commit, path=path, filesystem=filesystem),
            from_arrow_schema(pa.schema({"path": pa.string()})),
        ).coalesce(1).mapInArrow(
            partial(_rename, path=path, filesystem=filesystem),

>             from_arrow_schema(pa.schema({"path": pa.string()})),

Reviewer: What the purpose of committing and then renaming? I feel that the code would be more readable if we only preupload+commit (which is already complex logic in itself)

Author: Because when uploading I don't have the shard_idx information ^^'

Is it possible to do the CommitOperationAdd with different path_in_repo than the ones used for preupload_lfs ? This way now need for the rename step.

Author: Or maybe @anton-l if you know a way I can get the id of the partition passed in mapInArrow that I can use to get the shard_idx

Reviewer: Yes you can define whatever `path_in_repo` you want in the preupload step and change it afterwards before the `commit` step. It's hacky but the whole script seems to be for power users so :man_shrugging: In fact in the preupload step the `path_in_repo` is only used to determine if the file should be uploaded as LFS or not (rule is either "filesize > threshold" or "file extension in .gitattributes").

Author: I just simplified the code (significantly) using your trick, let me know what you think :)
